Statistics plugin improvements

phproberto opened the pull request:
## Description

This changes the JFormField to:
- Add a new `render()` function that can be used to render layouts inside other layouts. 
- Allow to easily override the layout include paths in any child field creating an override of the  `getLayoutPaths()` method.

Render layouts inside other layouts is used like:

``` php
<a href="http://joomla.org">Joomla</a>
<?php 
    // Render `sample.layout` layout inside this layout
    echo $field->render('sample.layout', $displayData);
?>
<p>Another paragraph</p>
```

You can see below how I used it to render a common `stats` layout that is renderer in both `message` and `field.data` layouts.

This PR also improves the statistics system plugin:
- Changed plugin fields prefix to plgsystemstats to avoid conflicts with any extension using `stats` prefix. Before: `type="stats.uniqueid"` after: `type="plgsystemstats.uniqueid"`
- Remove uniqueid field getInput() method and use the fields layout system so 100% of the plugin output is overridable.
- Create new field data that will be used to render the stats that are going to be sent in the plugin settings view.

![stats-params-data](https://cloud.githubusercontent.com/assets/1119272/11106927/4e93abc4-88d9-11e5-90cf-1b5ad6e6ed89.png)
## How to test
1. This changes the field layout system so you should test that fields that are currently using layouts work as expected. The only fields using it are `JFormFieldRadio` & `JFormFieldCheckboxes` so the best way to check is to edit the codemirror plugin that includes both fields in its settings.
2. Enable debug in statistics system plugin and test that the message selecting the user to configure statistics mode is shown. Also confirm that when you click in `Click here to see which information will be sent.` data is shown below.
3. Go to statistics system plugin settings and confirm that when you click in `Click here to see which information will be sent.` data appears below.

phproberto committed:
[imp] make JFormField::getRenderer() public and allow to override layout paths

phproberto committed:
[imp] plg_system_stats improvements

phproberto committed:
[imp] clean versions & translation strings for send data in plg_system_stats

phproberto commented:
I have also cleaned versions & added a language string for each data sent

![version-cleaned](https://cloud.githubusercontent.com/assets/1119272/11107319/1bf2fa22-88dd-11e5-8d93-0b896ce3f307.png)

zero-24 commented on plugins/system/stats/layouts/message.php:
```diff
@@ -23,12 +23,9 @@
-			<dd><?php echo $value; ?></dd>
-		<?php endforeach; ?>
-	</dl>
+	<?php
```
@phproberto can we move this here into one singel line?

`<?php echo $plugin->getRenderer('stats')->render(compact('statsData')); ?>`

phproberto replied:
When there is another layout being rendered I don't like the idea of using single lines because people can miss that there is another layout being rendered there. 

It's more an HTML block than a string text or a value being printed. That's why I did it like that.

zero-24 replied:
ok.

brianteeman commented:
I have tested this item :white_check_mark: successfully on 4dff4a2e398c2830b7a8dcf612837af2b9dd838b<br /><br />Great stuff - works as described<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8386">issues.joomla.org/joomla-cms/8386</a>.</sub>

wilsonge commented on plugins/system/stats/stats.php:
```diff
@@ -266,11 +280,11 @@ protected function getLayoutsPaths()
 	 * @since   3.5
 	 */
-	protected function getRenderer($layoutId = 'default')
+	public function getRenderer($layoutId = 'default')
```
Does this method need to be made public? I can understand the form field stuff. But you won't (I hope) be injecting the plugin object into a JLayout?

phproberto replied:
I am passing the plugin as layout parameter. Check:
https://github.com/joomla/joomla-cms/pull/8386/files#diff-15e6dcc8ba988545ca7bc176d5386f47R27

We need a way to render another layout with the plugin config (layout paths, etc.). Same requirement than for fields.

phproberto committed:
[imp] show error message when sending stats fails

phproberto commented:
I have updated the PR and added an extra check that will show the error message from the stats server if something when wrong (like data validations).

roland-d commented on plugins/system/stats/stats.xml:
```diff
@@ -19,11 +19,16 @@
 			<fieldset name="basic">
 				<field
+					name="data"
+					type="plgsystemstats.data"
```
@phproberto Is this new, types separated by a dot? If so, we should document that somewhere.

mbabker replied:
That has been part of the JForm API since it was merged into the old SVN trunk during development on 1.6.

phproberto replied:
yes. Not a lot of people knows it but it's the right way for extensions:
http://phproberto.com/en/26-joomla-form-fields-rules-right-way

roland-d replied:
Aha, thanks for the link @phproberto :)

wilsonge replied:
We already do this in the installer as well FWIW https://github.com/joomla/joomla-cms/commit/5680993ad06a2664550bdf99946c53e2138ccc37

roland-d commented:
I have tested this item :white_check_mark: successfully on e18b20c5aebb836def320669e9643727e346debc<br /><br />Works as described, the statistics plugin data is shown. The settings in CodeMirror keep working as they did before.

@phproberto This is more than just a statistics plugin improvement if I understand it correctly. You also changed the way JLayouts work by adding the getLayoutPaths() method by which form fields can override the list of paths to check the layout files. In the uniqueid.php and data.php we get duplicate code but I guess that can't be prevented the way it is setup now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8386">issues.joomla.org/joomla-cms/8386</a>.</sub>

phproberto committed:
[imp] keep getRenderer private and create render() public functions

phproberto commented:
I have updated the PR and added some changes:

* Avoid to make public getRenderer() methods in `JFormField` as suggested by @wilsonge and create a new public method called `render()`. That method will be used when you want to display a layout inside another field layout.
* I have created a new base field for the plugin fields to avoid the duplicated code reported by @roland-d 
* I have ensured that the data field only shows/hides the data attached to it.

phproberto commented:
And yes @roland-d I have modified the `JFormField` because I found that extensions (which is the case of this plugin) need to setup their own layout paths.

infograf768 commented:
Tested OK for codemirror and stats message.
Concerning that one I suggest to add some `<p>` in the layout as the link is not very visible on the blue back-ground.
Instead of
![screen shot 2015-12-03 at 12 17 29](https://cloud.githubusercontent.com/assets/869724/11559428/eb1a142a-99b8-11e5-97ca-2966825ffe60.png)
we would get
![screen shot 2015-12-03 at 12 21 33](https://cloud.githubusercontent.com/assets/869724/11559436/0b4e1fc0-99b9-11e5-8d34-f231715cae48.png)

Gerlof commented:
Am I rifght that this PR cannot be tested using Joomla! Patch Tester?
Clicking [Apply Patch] shows message "The file marked for modification does not exist: plugins/system/stats/layouts/message.php"

Bakual commented:
@gerlof You need to test this on staging. 3.5.0-beta may not work.

Gerlof commented:
@Bakual Joomla! Update shows "You are on the "Testing" update channel. This channel is designed for testing new releases and fixes in Joomla.
It is only intended for JBS (Joomla! Bug Squad™) members and others within the Joomla community who are testing. Do not use this setting on a production site.

You already have the latest Joomla version, 3.5.6-dev."

> How to change "3.5.6.-dev" to the right version? I think I have to change a setting in a db table, but which one to pick? ;)

Bakual commented:
I have no clue how you got onto a `3.5.6-dev`. That version doesn't exist.
I would just set up a new Joomla instance instead of trying to fix a broken one.

For staging, you can download https://github.com/joomla/joomla-cms/archive/staging.zip and install from that.

Gerlof commented:
@Bakual Yes, strange, isn't it? This happened after installing 3.5 Beta.

Gerlof commented:
@Bakual Note: wrong version number was caused by the file libraries/cms/version/version.php, from line 35:
>/**
	 * Maintenance version.
	 *
	 * @var    string
	 * @since  3.5
	 */
	const DEV_LEVEL = '6-dev';

Bakual commented:
Ah. @mbabker fixed that: https://github.com/joomla/joomla-cms/commit/5a20901097d8a375b7954f226038333c0ceba731

phproberto committed:
[imp] change copyright

phproberto commented:
Updated copyrights to 2016

infograf768 commented:
Any reason for using
`+ * @var   boolean  $spellchec       Spellcheck state for the form field.`
instead of 
`+ * @var   boolean  $spellcheck       Spellcheck state for the form field.`
?

dgt41 commented:
@infograf768 the way the vars get populated for the layout is by calling the parent class field and then merge any other, specific vars for the field. So this comment block is, for the most of it, generic...
Check https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/form/fields/checkboxes.php#L150-L167
and
https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/form/field.php#L952-L990

Fedik commented:
@dgt41 I think @infograf768 mean `$spellchec` vs `$spellcheck` :wink:

dgt41 commented:
@Fedik oops, that's a typo

phproberto committed:
[fix[ typo

phproberto commented:
Updated again. Please if this is not going to be merged save us time and close it.

andrepereiradasilva commented:
Nice work.
One question: Should the stats plugin autoload the language?
https://github.com/phproberto/joomla-cms/blob/43425a448031db5af1f95a09bd9bfcab1a7e4894/plugins/system/stats/stats.php#L52

infograf768 commented:
I have tested this item :white_check_mark: successfully on 7fb97f4cb8269e3bd5a7c7642fcc2af5a837462e<br /><br />One more tester needed.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8386">issues.joomla.org/joomla-cms/8386</a>.</sub>

Fedik commented:
all works if following the test instruction :wink: 

However I noticed issue, but not sure if it related to changes made in current pull request. 
When in the request permission I chose "allow" all work fine. But when I chose "once" or "never" then  the request permission message annoying me after each page refresh.

Gerlof commented:
I have tested this item :white_check_mark: successfully on 7fb97f4cb8269e3bd5a7c7642fcc2af5a837462e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8386">issues.joomla.org/joomla-cms/8386</a>.</sub>

MATsxm commented:
I have tested this item :white_check_mark: successfully on 7fb97f4cb8269e3bd5a7c7642fcc2af5a837462e<br /><br />tested again - all fine here
Thanks!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8386">issues.joomla.org/joomla-cms/8386</a>.</sub>
